recovery/elliptics_recovery/utils/misc.py

```diff
@@ -366,26 +368,36 @@ def run(self):
         while not self.complete.is_set():
             self.complete.wait()
 
-        speed = self.processed_keys / (time.time() - self.start_time)
-        self.stats.set_counter('recovery_speed', round(speed, 2))
-        self.stats.set_counter('recovers_in_progress', self.recovers_in_progress)
-        return self.result
+        self.pool.close()
+        self.pool.join()
 
-    def callback(self, result, stat):
-        self.run_one()
-        last = False
-        with self.lock:
-            self.result &= result
-            self.processed_keys += 1
-            self.recovers_in_progress -= 1
-            last = self.recovers_in_progress == 0
-
-        stat.apply(self.stats)
         speed = self.processed_keys / (time.time() - self.start_time)
         self.stats.set_counter('recovery_speed', round(speed, 2))
         self.stats.set_counter('recovers_in_progress', self.recovers_in_progress)
-        self.stats.counter('recovered_keys', 1 if result else -1)
-        self.ctx.stats.counter('recovered_keys', 1 if result else -1)
+        return self.result
 
-        if last:
-            self.complete.set()
+    def _callback(self, result, stat):
+        try:
+            self.run_one()
+
+            with self.lock:
+                self.result &= result
+                self.processed_keys += 1
+                self.recovers_in_progress -= 1
+                last = self.recovers_in_progress == 0
+
+            stat.apply(self.stats)
+            speed = self.processed_keys / (time.time() - self.start_time)
+            self.stats.set_counter('recovery_speed', round(speed, 2))
+            self.stats.set_counter('recovers_in_progress', self.recovers_in_progress)
+            self.stats.counter('recovered_keys', 1 if result else -1)
+            self.ctx.stats.counter('recovered_keys', 1 if result else -1)
+
+            if last:
+                self.complete.set()
+        except:
+            log.exception("Failed to perform callback")
```

Comment:
Any exception here will be skipped now without any any consequences. Is it correct?

Reply:
If we preserve old behavior, we can save exception object in any variable, and rethrow it after pool join.